ros2doctor/ros2doctor/api/package.py

```diff
@@ -0,0 +1,127 @@
+def _get_ros_packages_info(*, check=True) -> dict:
+    """
+    Return all current distro's packages info using rosdistro API.
+
+    :return: a dictionary contains package name, release, source and status
+    """
+    distro_name = os.environ.get('ROS_DISTRO')
+    if not distro_name:
+        if check:  # suppress warning when printing report
+            doctor_warn('ROS_DISTRO is not set.')
+        return
+    distro_name = distro_name.lower()
+    url = rosdistro.get_index_url()
+    if not url:
+        if check:  # suppress warning when printing report
+            doctor_warn('Unable to access ROSDISTRO_INDEX_URL or DEFAULT_INDEX_URL.')
+        return
+    i = rosdistro.get_index(url)
+    distro_data = rosdistro.get_distribution(i, distro_name).get_data()
+    return distro_data.get('repositories')
+
+
+def _get_local_package_version(package_name, *, check=True) -> str:
+    """
+    Return all locally installed packages info.
+
+    :param: str of package name
+    :return: a string contains local package version
+    """
+    try:
+        prefix = get_package_prefix(package_name)
+    except PackageNotFoundError:
+        if check:  # suppress warning when printing report
+            doctor_warn('Unable to find prefix for `%s` using index system.' % package_name)
+        return
+    xml_file = os.path.join(prefix, 'share', package_name, 'package.xml')
+    try:
+        root = ET.parse(xml_file).getroot()
+        for child in root:
+            if child.tag == 'version':
+                return child.text
+    except FileNotFoundError:
+        if check:  # suppress warning when printing report
+            doctor_warn('Unable to find package.xml of `%s`.' % package_name)
+
+
+class PackageCheck(DoctorCheck):
+    """Check local package versions against release versions on rosdistro."""
+
+    def category(self):
+        return 'package'
+
+    def check(self):
+        """Check packages within the directory where command is called."""
+        result = Result()
+        distro_packages_info = _get_ros_packages_info()
+        if not distro_packages_info:
+            result.add_error('ERROR: Unable to obtain current distro package information \
+                from rosdistro.')
+            return result
+        for package_name, info in distro_packages_info.items():
+            local_ver = _get_local_package_version(package_name)
+            try:
+                distro_release = info.get('release')
+                distro_ver = distro_release.get('version')
+            except AttributeError:
+                result.add_warning('Unable to obtain the rosdistro version \
+                    of `%s`.' % package_name)
+            if local_ver and distro_ver:
+                if distro_ver[:3] != local_ver[:3]:  # 0.8.0 vs. 0.8.0-1
+                    result.add_warning('`%s` has different local version %s from \
+                        required version %s' % (package_name, distro_ver, local_ver))
+        return result
```

Comment:
Instead of getting all packages and versions from rosdistro and then checking if they exist on the system I would suggest the opposite: collect all packages found and their versions and then compare them to the information from rosdistro.